### ISSUE TITLE
planning time is too long

### ISSUE THREAD

**Issue by lyg888** [prestodb/presto#16903]
I tail presto'log, and i see that planning time is too long,  so how make it better? 

8-e5cdd03ef3d2] :: elapsed 3554ms :: planning 2397ms :: scheduling 914ms :: running 206ms :: finishing 37ms :: begin 2021-10-22T16:44:20.447+08:00 :: end 2021-10-22T16:44:24.001+08:00
2021-10-22T16:44:31.573+0800    INFO    dispatcher-query-12     com.facebook.presto.event.QueryMonitor  TIMELINE: Query 20211022_084427_00020_buhv5 :: Transaction:[7a21b29f-5f5e-4325-8169-ec7aab41c43e] :: elapsed 3865ms :: planning 2630ms :: scheduling 1008ms :: running 183ms :: finishing 44ms :: begin 2021-10-22T16:44:27.513+08:00 :: end 2021-10-22T16:44:31.378+08:00
2021-10-22T16:44:59.421+0800    INFO    dispatcher-query-8      com.facebook.presto.event.QueryMonitor  TIMELINE: Query 20211022_084455_00021_buhv5 :: Transaction:[30530df9-0a22-4b78-8613-64e76133587c] :: elapsed 3397ms :: planning 2309ms :: scheduling 838ms :: running 220ms :: finishing 30ms :: begin 2021-10-22T16:44:55.842+08:00 :: end 2021-10-22T16:44:59.239+08:00

**Comment by yulongfufu** [comment]
I think it may be a simple sql according to the running time `220ms`.
Maybe if we can skip redundant optimizers for these small sql ?

Or the metadata access speed is your bottleneck...

**Comment by GithubZhitao** [comment]
"Skip small query" is a little tricky.  Maybe high concurrency services is not suitable for presto,  except that it can support HA .

**Comment by yulongfufu** [comment]
> Maybe high concurrency services is not suitable for presto

ohh, recently, we are thinking how to implement high concurrency in Presto.
> except that it can support HA .

In my opinion, I think HA and federation is two different things. We implemented HA to solve SPOF. It seems that what's your mean is federation ?

**Comment by GithubZhitao** [comment]
Yeah,  two problems exist simultaneously now.  HA is for cluster management or  Coordinator, while Parser and planning is 
 computing works, can it be separated when coding for HA? I mean they can be considered at the same time.  We can store some global info like cluster infos or global state machine in, like apache Doris using Berkeley DB. 

Again Two problems 
1.  SPOF
2.  Coordinator under too much pressure when large mount of small queries using too much CPU resource.

For 2.
Until now, we can't  let other machines share some works like SQLParser  PlanOptimization.  This is important when the cost of parsing and planning almost close to the query running on workers, which will support high-concurrency short queries. 
Actually, small queries occupy 50% ~ 80% .

Any discussing  for this feature that we can access ?   
We also have plan to implement it, or can wen open a issue for discussing further ?

**Comment by 7c00** [comment]
One of the most common causes of long planning time in our production environment is the long latency of api calls to hive metastore. You could diagnose the api call statistics (for example http://your-coordinator/v1/jmx/mbean/com.facebook.presto.hive.metastore.thrift:type=ThriftHiveMetastore,name=hive)

**Comment by yulongfufu** [comment]
> can it be separated when coding for HA

ok, I get your point.  Maybe it is a more complex method to consider both situations at the same time. 
Another idea: considering HDFS nameNode. HA and federation are separated into two steps. 
Maybe step by step is more acceptive, or maybe not   : )

> Any discussing for this feature that we can access ? 

email and wechat maybe more convenient. 

> We also have plan to implement it, or can wen open an issue for discussing further ?

Of course! But I just stand for my / our team points. Community's thoughts is not included.

cc @7c00

**Comment by GithubZhitao** [comment]
@yulongfufu. 
hi,The feature will come soon. 

https://github.com/prestodb/presto/issues/15453#issuecomment-1039802250
